# Hypothesis Testing 
---
#### Problem Statement 

* An online bookstore is looking to optimize its website design to improve user engagement and ultimately increase book purchases. The website currently offers two themes for its users: “Light Theme” and “Dark Theme.” The bookstore’s data science team wants to conduct an Hypothesis Testing and A/B Testing  experiment to determine which theme leads to better user engagement and higher conversion rates for book purchases.

#### Goal

* Task is to identify which theme, Light Theme or Dark Theme, yields better user engagement, purchases and conversion rates. You need to determine if there is a statistically significant difference in the key metrics between the two themes.


### Import Required Libraries 

In [1]:
import numpy as np 
import pandas as pd 
from scipy.stats import ttest_ind

In [4]:
### Load CSV file
df= pd.read_csv("C:\\Users\\Daman\\Downloads\\website_ab_test.csv")
df.head()

Theme  Click Through Rate  Conversion Rate  Bounce Rate  \
0  Light Theme            0.054920         0.282367     0.405085   
1  Light Theme            0.113932         0.032973     0.732759   
2   Dark Theme            0.323352         0.178763     0.296543   
3  Light Theme            0.485836         0.325225     0.245001   
4  Light Theme            0.034783         0.196766     0.765100   

   Scroll_Depth  Age   Location  Session_Duration Purchases Added_to_Cart  
0     72.489458   25    Chennai              1535        No           Yes  
1     61.858568   19       Pune               303        No           Yes  
2     45.737376   47    Chennai               563       Yes           Yes  
3     76.305298   58       Pune               385       Yes            No  
4     48.927407   25  New Delhi              1437        No            No

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Theme               1000 non-null   object 
 1   Click Through Rate  1000 non-null   float64
 2   Conversion Rate     1000 non-null   float64
 3   Bounce Rate         1000 non-null   float64
 4   Scroll_Depth        1000 non-null   float64
 5   Age                 1000 non-null   int64  
 6   Location            1000 non-null   object 
 7   Session_Duration    1000 non-null   int64  
 8   Purchases           1000 non-null   object 
 9   Added_to_Cart       1000 non-null   object 
dtypes: float64(4), int64(2), object(4)
memory usage: 78.3+ KB


* There are no null values in entire dataset 

In [11]:
df.describe().T

count        mean         std        min         25%  \
Click Through Rate  1000.0    0.256048    0.139265   0.010767    0.140794   
Conversion Rate     1000.0    0.253312    0.139092   0.010881    0.131564   
Bounce Rate         1000.0    0.505758    0.172195   0.200720    0.353609   
Scroll_Depth        1000.0   50.319494   16.895269  20.011738   35.655167   
Age                 1000.0   41.528000   14.114334  18.000000   29.000000   
Session_Duration    1000.0  924.999000  508.231723  38.000000  466.500000   

                           50%          75%          max  
Click Through Rate    0.253715     0.370674     0.499989  
Conversion Rate       0.252823     0.373040     0.498916  
Bounce Rate           0.514049     0.648557     0.799658  
Scroll_Depth         51.130712    64.666258    79.997108  
Age                  42.000000    54.000000    65.000000  
Session_Duration    931.000000  1375.250000  1797.000000

* **Click Through Rate**: Ranges from about 0.01 to 0.50 with a mean of approximately 0.26.

* **Conversion Rate**: Also ranges from about 0.01 to 0.50 with a mean close to the Click Through Rate, approximately 0.25.

* **Bounce Rate**: Varies between 0.20 and 0.80, with a mean around 0.51.

* **Scroll Depth**: Shows a spread from 20.01 to nearly 80, with a mean of 50.32.

* **Age**: The age of users ranges from 18 to 65 years, with a mean age of about 41.5 years.

* **Session Duration**: This varies widely from 38 seconds to nearly 1800 seconds (30 minutes), with a mean session duration of approximately 925 seconds (about 15 minutes).

#### Comparing the performance of both themes based on the provided metrics

In [14]:
### Remove categorical and string type of columns from dataset
column_names = df.columns.tolist()
print(column_names)

['Theme', 'Click Through Rate', 'Conversion Rate', 'Bounce Rate', 'Scroll_Depth', 'Age', 'Location', 'Session_Duration', 'Purchases', 'Added_to_Cart']


In [16]:
drop_columns = ['Location','Purchases', 'Added_to_Cart']
test_df = df.drop(columns=drop_columns)

In [21]:
## calcuate mean for all the metrics by themes 
theme_perfomance = test_df.groupby('Theme').mean()
theme_perfomance = pd.DataFrame(theme_perfomance)

In [22]:
theme_perfomance

Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
Theme                                                                         
Dark Theme             0.264501         0.251282     0.512115     49.926404   
Light Theme            0.247109         0.255459     0.499035     50.735232   

                   Age  Session_Duration  
Theme                                     
Dark Theme   41.332685        919.482490  
Light Theme  41.734568        930.833333

* **Click Through Rate (CTR)**: Dark Theme has a slightly higher CTR (0.2645) compared to Light Theme (0.2471). This suggests that users might find the Dark Theme more engaging or appealing in terms of click interaction.<br>
<br>
* **Conversion Rate**: The conversion rate is similar for both themes, with Light Theme slightly higher at 0.2555 compared to Dark Theme at 0.2513. This indicates that both themes are equally effective in converting users.<br>
<br>
* **Bounce Rate**: Dark Theme has a higher bounce rate (0.5121) compared to Light Theme (0.4990). This suggests that users might find the Dark Theme less engaging or informative, leading to more immediate exits from the page.<br>
<br>
* **Scroll Depth**: The average scroll depth is slightly higher for Light Theme (50.74) compared to Dark Theme (49.93). This could indicate that users spend slightly more time scrolling on pages with the Light Theme.<br>
<br>
* **Session Duration**: The average session duration is similar for both themes, with Light Theme having a slightly higher duration (930.83) compared to Dark Theme (919.48). This suggests that users might spend slightly more time on pages with the Light Theme.

#### Defind null H0 and Ha Hypothesis 
---

We’ll use a significance level (alpha) of 0.05 for our hypothesis testing. It means we’ll consider a result statistically significant if the p-value from our test is less than 0.05.

start with hypothesis testing based on the Conversion Rate ,Click Through Rate(CTR), Bounce Rate, Scroll Depth, Session Duration respectively between the Light Theme and Dark Theme. 


* **Null Hypothesis (H0​)**: There is no difference in performance  between the Light Theme and Dark Theme.<br>
<br>
* **Alternative Hypothesis (Ha​)**: There is a difference in performance  between the Light Theme and Dark Theme.<br>
<br>

In [23]:
###lets perform t-test to validate null hypothesis 

# extracting conversion rates for both themes
conversion_rates_light = df[df['Theme'] == 'Light Theme']['Conversion Rate']
conversion_rates_dark = df[df['Theme'] == 'Dark Theme']['Conversion Rate']

# performing a two-sample t-test
t_stat, p_value = ttest_ind(conversion_rates_light, conversion_rates_dark, equal_var=False)

print("t-score"t_stat, p_value

(0.4748494462782632, 0.6349982678451778)

* The result of the two-sample t-test gives a p-value of approximately 0.635(**63.5%**). Since this p-value is much greater than our significance level of 0.05, we do not have enough evidence to reject the null hypothesis. Therefore, we conclude that there is no statistically significant difference in Conversion Rates between the Light Theme and Dark Theme based on the data provided.

In [24]:
from scipy.stats import ttest_ind

# Define the metrics for which you want to compare conversion rates
metrics = ['Click Through Rate', 'Session_Duration', 'Bounce Rate', 'Scroll_Depth']

# Initialize an empty dictionary to store p-values for each metric
p_values = {}

# Iterate over each metric
for metric in metrics:
    # Extract conversion rates for both themes
    conversion_rates_light = df[df['Theme'] == 'Light Theme'][metric]
    conversion_rates_dark = df[df['Theme'] == 'Dark Theme'][metric]
    
    # Perform a two-sample t-test
    t_stat, p_value = ttest_ind(conversion_rates_light, conversion_rates_dark, equal_var=False)
    
    # Store the p-value for the metric
    p_values[metric] = p_value
    

# Display the p-values for each metric
for metric, p_value in p_values.items():
    print(f'P-value for {metric}: {p_value}')


P-value for Click Through Rate: 0.04818435371010704
P-value for Session_Duration: 0.7242292543366338
P-value for Bounce Rate: 0.229692077505148
P-value for Scroll_Depth: 0.4496919249484911


In [26]:
import pandas as pd
from scipy.stats import ttest_ind

# Define the metrics for which you want to compare conversion rates
metrics = ['Click Through Rate', 'Session_Duration', 'Bounce Rate', 'Scroll_Depth']

# Initialize empty lists to store t-values and p-values
t_values = []
p_values = []

# Iterate over each metric
for metric in metrics:
    # Extract conversion rates for both themes
    conversion_rates_light = df[df['Theme'] == 'Light Theme'][metric]
    conversion_rates_dark = df[df['Theme'] == 'Dark Theme'][metric]
    
    # Perform a two-sample t-test
    t_stat, p_value = ttest_ind(conversion_rates_light, conversion_rates_dark, equal_var=False)
    
    # Append the t-value and p-value to the respective lists
    t_values.append(t_stat)
    p_values.append(p_value)

# Create a DataFrame from the t-values and p-values
results_df = pd.DataFrame({'Metric': metrics, 't-value': t_values, 'p-value': p_values})

results_df


Metric   t-value   p-value
0  Click Through Rate -1.978171  0.048184
1    Session_Duration  0.352912  0.724229
2         Bounce Rate -1.201888  0.229692
3        Scroll_Depth  0.756228  0.449692

* **Click Through Rate**: The test reveals a statistically significant difference, with the Dark Theme likely performing better (P-Value = 0.048).<br>
<br>
* **Conversion Rate**: No statistically significant difference was found (P-Value = 0.635).<br>
<br>
* **Bounce Rate**: There’s no statistically significant difference in Bounce Rates between the themes (P-Value = 0.230).<br>
<br>
* **Scroll Depth**: Similarly, no statistically significant difference is observed in Scroll Depths (P-Value = 0.450).<br>
<br>
---

### Summary 

while the two themes perform similarly across most metrics, the Dark Theme has a slight edge in terms of engaging users to click through. For other key performance indicators like Conversion Rate, Bounce Rate, and Scroll Depth, the choice between a Light Theme and a Dark Theme does not significantly affect user behaviour according to the data provided